cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java

```diff
@@ -0,0 +1,206 @@
+        v3MetricsTable.increment(rowIncrements);
+      }
+      // delete entries from old table
+      // break if we have exceeded max records to scan
+      recordsScanned++;
+    }
+    scanner.close();
+    LOG.trace("Migrated {} records from the metrics table {}", recordsScanned, v2MetricsTable);
+  }
+
+  private byte[][] getByteArrayFromSets(Set<byte[]> incrementSet, Set<byte[]> gaugeSet) {
+    byte [][] deletes = new byte[incrementSet.size() + gaugeSet.size()][];
+    int index = 0;
+    for (byte[] column : incrementSet) {
+      deletes[index++] = column;
+    }
+    for (byte[] column : gaugeSet) {
+      deletes[index++] = column;
+    }
+    return deletes;
+  }
+
+  private TableId getV2MetricsTableName(int resolution) {
+    String v2TableName = cConf.get(Constants.Metrics.METRICS_TABLE_PREFIX,
+                                   Constants.Metrics.DEFAULT_METRIC_TABLE_PREFIX) + ".ts." + resolution;
+    return TableId.from(NamespaceId.SYSTEM.getNamespace(), v2TableName);
+  }
+
+  /**
+   * check if v2 metrics table exists for the resolution
+   * @param tableUtil hbase table util
+   * @param resolution resolution of the metrics table to check
+   * @return true if table exists; false otherwise
+   */
+  public boolean v2MetricsTableExists(HBaseTableUtil tableUtil, int resolution) {
+    TableId tableId = getV2MetricsTableName(resolution);
+    try {
+      try (HBaseAdmin admin = new HBaseAdmin(hConf)) {
+        TableId hBaseTableId =
+          tableUtil.createHTableId(new NamespaceId(tableId.getNamespace()), tableId.getTableName());
+        boolean doesExist  = tableUtil.tableExists(admin, hBaseTableId);
+        LOG.trace("Table {} exists ? : {}", hBaseTableId.getTableName(), doesExist);
+        return doesExist;
+      }
+    } catch (IOException e) {
+      LOG.warn("Exception while checking table exists", e);
+    }
+    return false;
+  }
+
+  /**
+   * delete v2 metrics table for the resolution
+   * @param ddlExecutorFactory HBaseDDLExecutorFactory
+   * @param tableUtil hbase table util
+   * @param resolution resolution of metrics table to delete
+   * @return true if deletion is successful; false otherwise
+   */
+  public boolean deleteV2MetricsTable(HBaseDDLExecutorFactory ddlExecutorFactory,
+                                      HBaseTableUtil tableUtil, int resolution) {
+    TableId tableId = getV2MetricsTableName(resolution);
+    LOG.trace("Looking to delete table {}", tableId);
+    try {
+      try (HBaseDDLExecutor ddlExecutor = ddlExecutorFactory.get(); HBaseAdmin admin = new HBaseAdmin(hConf)) {
+        TableId hBaseTableId =
+          tableUtil.createHTableId(new NamespaceId(tableId.getNamespace()), tableId.getTableName());
+        if (tableUtil.tableExists(admin, hBaseTableId)) {
+          LOG.trace("Found table {}, going to delete", hBaseTableId);
+          tableUtil.dropTable(ddlExecutor, hBaseTableId);
+          LOG.trace("Deleted table {}", hBaseTableId);
```

Comment:
what happens if a query is currently reading that table? Do you need to change something in the read path to catch the exception and retry?